File: src/com/systemvi/examples/physics2d/Physics.java
package com.systemvi.examples.physics2d;

import com.systemvi.engine.application.Application;
import com.systemvi.engine.camera.Camera;
import com.systemvi.engine.renderers.ShapeRenderer;
import com.systemvi.engine.window.Window;
import org.jbox2d.collision.shapes.PolygonShape;
import org.jbox2d.collision.shapes.Shape;
import org.jbox2d.common.Vec2;
import org.jbox2d.dynamics.*;
import org.joml.Vector4f;

import java.util.ArrayList;

import static org.lwjgl.opengl.GL33.*;

public class Physics extends Application {
    public Physics(int openglVersionMajor, int openglVersionMinor, int targetFPS) {
        super(openglVersionMajor, openglVersionMinor, targetFPS);
    }

    Window window;
    World world;
    ShapeRenderer renderer;
    Camera camera;
    float x,y;
    ArrayList<Body> bodies;

    @Override
    public void setup() {
        window=new Window(800,600,"Box2d demo");
        world=new World(new Vec2(0,600));
        renderer=new ShapeRenderer();
        camera=new Camera();
        camera.setPosition(400,300,0);
        camera.setScreenSize(800,600);
        camera.setScale(1,-1,1);
        camera.update();
        renderer.setCamera(camera);
        window.addOnMouseMoveListener((x1, y1) -> {
            x=(int)x1;
            y=(int)y1;
        });
        bodies=new ArrayList<>();

        window.addOnMouseDownListener((button, mods) -> {
            BodyDef bodyDef=new BodyDef();
            bodyDef.position.set(x,y);
            bodyDef.type= BodyType.DYNAMIC;
            bodyDef.linearDamping=0;
            FixtureDef fixtureDef=new FixtureDef();
            fixtureDef.density=1;
            fixtureDef.restitution=0.5f;
            fixtureDef.friction=0.7f;
            PolygonShape shape=new PolygonShape();
            shape.setAsBox(20f,20f);

            fixtureDef.shape=shape;

            Body body=world.createBody(bodyDef);
            body.createFixture(fixtureDef);
            bodies.add(body);
        });
        createWalls();
    }

    @Override
    public void loop(float delta) {
        if(window.shouldClose())close();
        glClearColor(0,0,0,1);
        glClear(GL_COLOR_BUFFER_BIT);
        //input
        window.pollEvents();
        //update
        world.step(delta,10,10);
        //draw
//        renderer.rect();
        for(int i=0;i<bodies.size();i++){
            Body body=bodies.get(i);
            float x=body.getPosition().x;
            float y=body.getPosition().y;
            float a=body.getAngle();
            renderer.rect(x-20,y-20,40,40,new Vector4f(1),a);
        }
        renderer.flush();

        window.swapBuffers();
    }

    public void createWalls(){
        BodyDef bodyDef=new BodyDef();
        bodyDef.position.set(400f,560f);
        bodyDef.type= BodyType.STATIC;

        FixtureDef fixtureDef=new FixtureDef();
        fixtureDef.density=1;
        fixtureDef.restitution=0.5f;
        fixtureDef.friction=0.7f;
        PolygonShape shape=new PolygonShape();
        shape.setAsBox(400f,21f);

        fixtureDef.shape=shape;

        Body body=world.createBody(bodyDef);
        body.createFixture(fixtureDef);
        bodies.add(body);
    }
}

File: examples/physics2d/src/scala/com/systemvi/physics2d/Physics.scala
package com.systemvi.physics2d;

import com.systemvi.engine.application.{Application, Game}
import com.systemvi.engine.camera.{Camera, Camera3}
import com.systemvi.engine.physics2d.PhysicsBody
import com.systemvi.engine.renderers.ShapeRenderer
import com.systemvi.engine.ui.utils.data.Colors
import com.systemvi.engine.utils.Utils
import com.systemvi.engine.utils.Utils.Buffer
import com.systemvi.engine.window.Window
import org.jbox2d.collision.shapes.PolygonShape
import org.jbox2d.common.Vec2
import org.jbox2d.dynamics.*
import org.joml.{Vector2f, Vector4f}

import java.util.ArrayList
import org.lwjgl.opengl.{GL11, GL33}

case class PhysicsBody(body:Body,size:Vector2f)

object Physics extends Game(3, 3, 60, 800, 600, "Physics") {

  var world: World = null
  var renderer: ShapeRenderer = null
  var camera: Camera3 = null
  var mouse = Vector2f()
  var bodies: Array[PhysicsBody] = Array()
  val size: Float = 20

  override def setup(window: Window): Unit = {
    val width = window.getWidth.toFloat
    val height = window.getHeight.toFloat
    world = World(new Vec2(0, 1000))
    renderer = ShapeRenderer()
    camera = Camera3.builder2d()
      .position(width / 2, height / 2)
      .size(width, height)
      .build()

    camera.update()
    renderer.setView(camera.view)
    renderer.setProjection(camera.projection)

    createWalls()
  }

  override def loop(delta: Float): Unit = {
    val window = getWindow
    if (window.shouldClose()) close()
    Utils.clear(Colors.black, Buffer.COLOR_BUFFER)
    //input
    window.pollEvents()
    //update
    world.step(delta, 10, 10)
    //draw
    for (element <- bodies) {
      val x = element.body.getPosition.x
      val y = element.body.getPosition.y
      val a = element.body.getAngle
      renderer.rect(
          x - element.size.x/2f,
          y - element.size.y/2f,
          element.size.x,
          element.size.y,
          Colors.red400, a
      )
    }
    renderer.flush()
  }

  def createWalls(): Unit = {
    addBody(0,300,20,600,BodyType.STATIC)
    addBody(780,300,20,600,BodyType.STATIC)
    addBody(400,0,800,20,BodyType.STATIC)
    addBody(400,580,800,20,BodyType.STATIC)
  }

  override def mouseDown(button: Int, mods: Int, x: Double, y: Double): Boolean = {
    addBody(x.toFloat,y.toFloat,20,20)
    true
  }

  def addBody(x:Float,y:Float,width:Float,height:Float,`type`:BodyType=BodyType.DYNAMIC):Unit = {
      val bodyDef = BodyDef()
      bodyDef.position.set(x, y)
      bodyDef.`type` = `type`
      bodyDef.linearDamping = 0
      bodyDef.angularDamping = 0

      val fixtureDef = FixtureDef()
      fixtureDef.density = 1
      fixtureDef.restitution = 0.5f
      fixtureDef.friction = 0.7f
      val shape = PolygonShape()
      shape.setAsBox(width/2, height/2)

      fixtureDef.shape = shape

      val body = world.createBody(bodyDef)
      body.createFixture(fixtureDef)
      bodies=bodies :+ PhysicsBody(body, Vector2f(width,height))
  }

  override def resize(width: Int, height: Int): Boolean = {
    camera.position.set(width / 2f, height / 2f, 0)
    camera.orthographic(-width / 2f, width / 2f, -height / 2f, height / 2f, 0, 1)
    camera.update()
    true
  }

  override def mouseMove(x: Double, y: Double): Boolean = {
    mouse.set(x, y)
    true
  }
}
